File: src/com/reus/leetcode/JumpGame.java
package com.reus.leetcode;

public class JumpGame {
    public boolean canJump(int[] nums) {
        int maxReachable = -1;
        for (int i = 0; i < nums.length; i++) {
            if (i > maxReachable && i != 0) {
                return false;
            }
            maxReachable = Math.max(i + nums[i], maxReachable);
            if (maxReachable >= nums.length - 1) {
                return true;
            }
        }
        return false;
    }
}

File: src/com/reus/leetcode/google/arrays/JumpGame.kt
package com.reus.leetcode.google.arrays

/**
 * https://leetcode.com/problems/jump-game
 */
class JumpGame {
    fun canJump(nums: IntArray): Boolean {
        val reachability = BooleanArray(nums.size) { false }
        reachability[reachability.size - 1] = true
        var curTargetIdx = reachability.size - 1
        for (i in reachability.size - 2 downTo 0) {
            if (i + nums[i] >= curTargetIdx) {
                reachability[i] = true
                curTargetIdx = i
            }
        }
        return reachability[0]
    }
}

fun main() {
    JumpGame().canJump(intArrayOf(3,2,1,0,4))
}